sklearn_numba_dpex/kmeans/kernels.py

```diff
@@ -0,0 +1,745 @@
+
+        n_samples = v.shape[0]
+
+        shm = dpex.local.array(work_group_size, dtype=dtype)
+
+        first_sample_idx = group_id * work_group_size * 2
+        augend_idx = first_sample_idx + local_work_id
+        addend_idx = first_sample_idx + work_group_size + local_work_id
+
+        if augend_idx >= n_samples:
+            shm[local_work_id] = zero
+        elif addend_idx >= n_samples:
+            shm[local_work_id] = v[augend_idx]
+        else:
+            shm[local_work_id] = v[augend_idx] + v[addend_idx]
+
+        dpex.barrier()
+        current_size = work_group_size
+        for i in range(local_nb_iterations):
+            current_size = current_size // 2
+            if local_work_id < current_size:
+                shm[local_work_id] += shm[local_work_id + current_size]
+
+            dpex.barrier()
+
+        if first_work_id:
+            w[group_id] = shm[0]
+
+    _steps_data = []
+    n_groups = n_samples
+    # TODO: at some point, the cost of scheduling the kernel is more than the cost of
+    # running the reduction iteration. At this point the loop should stop and then a
+    # single work item should iterates one time on the remaining values to finish the
+    # reduction.
+    while n_groups > 1:
+        n_groups = math.ceil(n_groups / (2 * work_group_size))
+        global_size = n_groups * work_group_size
+        _steps_data.append(
+            (
+                sum_reduction_kernel[global_size, work_group_size],
+                dpctl.tensor.empty(n_groups, dtype=dtype, device=device),
+            )
+        )
+
+    def sum_reduction(v):
+        for sum_fn, w in _steps_data:
+            sum_fn(v, w)
+            v = w
+        return v
+
+    return sum_reduction
+
+
+@lru_cache
+def make_sum_reduction_2dim_kernel(n_samples, n_features, work_group_size, dtype):
+    global_size = math.ceil(n_samples / work_group_size) * work_group_size
+    zero = dtype(0.0)
+
+    @dpex.kernel
+    def sum_reduction_kernel(v, w):
+        sample_idx = dpex.get_global_id(0)
+        if sample_idx >= n_samples:
+            return
+        tmp = zero
+        for feature_idx in range(n_features):
+            tmp += v[feature_idx, sample_idx]
+        w[sample_idx] = tmp
+
+    return sum_reduction_kernel[global_size, work_group_size]
+
+
+@lru_cache
+def make_sum_reduction_3dim_kernel(dim0, dim1, dim2, work_group_size, dtype):
+    nb_groups_for_dim2 = math.ceil(dim2 / work_group_size)
+    nb_threads_for_dim2 = nb_groups_for_dim2 * work_group_size
+    global_size = nb_threads_for_dim2 * dim1
+    zero = dtype(0.0)
+
+    @dpex.kernel
+    def sum_reduction_kernel(v, w):
+        group = dpex.get_group_id(0)
+        thread = dpex.get_local_id(0)
+        d1 = group // nb_groups_for_dim2
+        d2 = ((group % nb_groups_for_dim2) * work_group_size) + thread
+        if d2 >= dim2:
+            return
+        tmp = zero
+        for d in range(dim0):
+            tmp += v[d, d1, d2]
+        w[d1, d2] = tmp
```

Comment:
Wouldn't it be more natural and more efficient to directly use a 2D global id instead of manually defining `d1` and `d2` via integer division / modulo?

Reply:
Absolutely, in this particular case where the input is 3d the 2d grid sounds very much better, very good point !!

Reply:
But one need to be extra careful that the work group dimension that spans contiguous values of the input data is itself a multiple of the preferred_work_group_size_multiple, and also that work items along this dimensions should also be "contiguous" and span "warps".

It would need a bit more of testing to ensure that those considerations really matter and watch out the performance.

Reply:
ok for adding a TODO comment in this PR + an issue in the tracker for a follow-up PR later.

Reply:
Also anther strategy that could be benchmarked is to use [`numba.njit` with `dpnp`](https://intelpython.github.io/numba-dpex/latest/developer_guides/dpnp_integration.html) over writing from scratch with `dpex`